k2/csrc/cuda/debug.cuh

```diff
@@ -0,0 +1,169 @@
+// k2/csrc/cuda/errors.h
+
+// Copyright (c) 2020, Xiaomi Corporation ( authors: Meixu Song )
+
+// See LICENSE for clarification regarding multiple authors
+
+#ifndef K2_CSRC_CUDA_ERRORS_H_
+#define K2_CSRC_CUDA_ERRORS_H_
+
+#include <cstdio>
+
+#include "k2/csrc/cuda/arch.h"
+
+namespace k2 {
+
+// In debug mode and include this header,
+// define K2_MAKE_ERROR_CHECK to turn on error checking
+#if (!defined(NDEBUG) && !defined(K2_MAKE_ERROR_CHECK))
+  #define K2_MAKE_ERROR_CHECK
+#endif
+
+/**
+ * \brief A static assertion
+ * \param exp the compile-time boolean expression that must be true
+ * \param msg an error message if exp is false
+ *
+ * `static_assert` is supported by both of host and device.
+ */
+#ifndef K2_STATIC_ASSERT
+  #define K2_STATIC_ASSERT(exp, msg) static_assert(exp, msg)
+#endif
+
+/**
+ * \brief This is a error checking function, with context information.
+ * If K2_MAKE_ERROR_CHECK is defined and error is not cudaSuccess,
+ * the corresponding error message is printed to:
+ *   - host: stderr
+ *   - device: stdout in device
+ * along with the supplied source context.
+ * It's used to make other macros convenient.
+ *
+ * \return The CUDA error.
+ */
+__host__ __device__ __forceinline__ cudaError_t K2_CUDA_DEBUG(
```

Comment:
The function `K2_CUDA_DEBUG` is already inside the namespace `k2`, so there is
no need to put a `k2` in it.

`K2_CUDA_DEBUG` should be a macro name, not a function name.

Reply:
> K2_CUDA_DEBUG should be a macro name, not a function name

I agree with @csukuangfj  here , it's better if we can change `K2_CUDA_DEBUG` to follow name convention for functions (as it's a function instead of macro), e.g. `CudaDebug`, as you have defined macros below (e.g. `K2_CUDA_CHECK_ERROR`) to provide `__FILE__` and `__LINE__`. I suppose function `K2_CUDA_DEBUG` will not be called directly by users.
 
BTW, I'm in favour of macros with `K2_` prefix.

Reply:
```cpp
void CudaDebug(....) {....}

#define K2_CUDA_DEBUG(...) CudaDebug(...)
```

This is what I meant.

It's fine to have `K2` in the macro, but I don't think it is needed in a function which is inside the namespace k2.

Reply:
> The function `K2_CUDA_DEBUG` is already inside the namespace `k2`, so there is
> no need to put a `k2` in it.

I am not very certain. But if it gets called out the `k2` namespace.. as it's an inline function.

Reply:
> > K2_CUDA_DEBUG should be a macro name, not a function name
> 
> I agree with @csukuangfj here , it's better if we can change `K2_CUDA_DEBUG` to follow name convention for functions (as it's a function instead of macro), e.g. `CudaDebug`, as you have defined macros below (e.g. `K2_CUDA_CHECK_ERROR`) to provide `__FILE__` and `__LINE__`. I suppose function `K2_CUDA_DEBUG` will not be called directly by users.
> 
> BTW, I'm in favour of macros with `K2_` prefix.

The inline function name changed from "K2_CUDA_DEBUG" to "_K2DebugLog", "_" indicates it's an inner func.

Reply:
All functions in the namespace k2, except this one, do not contain `k2`, so what is the reason for adding k2 to this function?

Reply:
> All functions in the namespace k2, except this one, do not contain `k2`, so what is the reason for adding k2 to this function?

I know what you mean, but you may not try to get my point. For example:
```cpp
#include <iostream>

namespace k2 {
  inline void Print() {
    std::cout << "Hello World" << std::endl;
  }

#define PP Print()
}

int main () {
  PP;
  return 0;
}
```

compile and get error:
```bash
note: 'k2::Print' declared here
  inline void Print() {
              ^
1 error generated. 
```

Reply:
Could you use

```cpp
#define PP k2::Print()
```

The compiler gives a hint that you can use `k2::Print` in the macro.